File: app/src/main/java/org/ametro/model/serialization/CommonTypes.java
package org.ametro.model.serialization;

import android.graphics.Color;
import android.util.Log;

import com.fasterxml.jackson.databind.JsonNode;

import org.ametro.app.Constants;
import org.ametro.model.entities.MapPoint;
import org.ametro.model.entities.MapRect;
import org.ametro.utils.StringUtils;

public class CommonTypes {

    public final static int DEFAULT_COLOR = 0x0000FF;
    public final static int DEFAULT_LABEL_BG_COLOR = 0x0000FF;

    public static String[] asStringArray(JsonNode node) {
        String[] array = new String[node.size()];
        for (int i = 0; i < node.size(); i++) {
            array[i] = node.get(i).asText();
        }
        return array;
    }

    public static int asColor(JsonNode node, int defaultColor) {
        if(node == null || node.isNull()){
            return defaultColor;
        }

        String text = node.asText();
        if(StringUtils.isNullOrEmpty(text))
            return defaultColor;

        if(text.equals("-1")){ // TODO: fix on server side
            return Color.parseColor("#FFFFFF");
        }
        try {
            return Color.parseColor('#' + text);
        }catch (Exception ex){
            Log.e(Constants.LOG, "Invalid color [" + text + "]");
            return defaultColor;
        }
    }

    public static MapRect asRect(JsonNode node) {
        if(node == null || node.isNull()){
            return null;
        }
        return new MapRect(
                node.get(0).asInt(),
                node.get(1).asInt(),
                node.get(2).asInt(),
                node.get(3).asInt());
    }

    public static MapPoint asPoint(JsonNode node) {
        if(node == null || node.isNull()){
            return null;
        }
        return new MapPoint(
                (float)node.get(0).asDouble(),
                (float)node.get(1).asDouble());
    }

    public static MapPoint[] asPointArray(JsonNode arrayNode) {
        if(arrayNode == null || arrayNode.isNull()){
            return new MapPoint[0];
        }
        MapPoint[] array = new MapPoint[arrayNode.size()];
        for (int i = 0; i < arrayNode.size(); i++) {
            array[i] = asPoint(arrayNode.get(i));
        }
        return array;
    }

    public static int[] asIntArray(JsonNode arrayNode) {
        if(arrayNode == null || arrayNode.isNull()){
            return new int[0];
        }
        int[] array = new int[arrayNode.size()];
        for (int i = 0; i < arrayNode.size(); i++) {
            array[i] = arrayNode.get(i).asInt();
        }
        return array;
    }
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/model/serialization/CommonTypes.kt
package io.github.romangolovanov.apps.ametro.model.serialization

import android.graphics.Color
import android.util.Log

import com.fasterxml.jackson.databind.JsonNode

import io.github.romangolovanov.apps.ametro.app.Constants
import io.github.romangolovanov.apps.ametro.model.entities.MapPoint
import io.github.romangolovanov.apps.ametro.model.entities.MapRect

object CommonTypes {
    val DEFAULT_COLOR: Int = 0x0000FF
    val DEFAULT_LABEL_BG_COLOR: Int = 0x0000FF

    fun asStringArray(node: JsonNode): Array<String> {
        return Array(node.size()) { i -> node.get(i).asText() }
    }

    fun asColor(node: JsonNode?, defaultColor: Int): Int {
        if (node == null || node.isNull) return defaultColor
        val text = node.asText()
        if (text.isNullOrBlank()) return defaultColor
        if (text == "-1") return Color.parseColor("#FFFFFF")
        return try {
            Color.parseColor('#' + text)
        } catch (ex: Exception) {
            Log.e(Constants.LOG, "Invalid color [$text]")
            defaultColor
        }
    }

    fun asRect(node: JsonNode?): MapRect? {
        if (node == null || node.isNull) return null
        return MapRect(
            node.get(0).asInt(),
            node.get(1).asInt(),
            node.get(2).asInt(),
            node.get(3).asInt()
        )
    }

    fun asPoint(node: JsonNode?): MapPoint? {
        if (node == null || node.isNull) return null
        return MapPoint(
            node.get(0).asDouble().toFloat(),
            node.get(1).asDouble().toFloat()
        )
    }

    fun asPointArray(arrayNode: JsonNode?): Array<MapPoint> {
        if (arrayNode == null || arrayNode.isNull) return emptyArray()
        return Array(arrayNode.size()) { i ->
            asPoint(arrayNode.get(i)) ?: error("Unexpected null point at index $i")
        }
    }

    fun asIntArray(arrayNode: JsonNode?): IntArray {
        if (arrayNode == null || arrayNode.isNull) return IntArray(0)
        return IntArray(arrayNode.size()) { i -> arrayNode.get(i).asInt() }
    }
}
